centrallix-sysdoc/OSDriver_Authoring.md

```diff
@@ -115,497 +170,1032 @@ License:  Copyright (C) 2001-2011 LightSys Technology Services. See LICENSE.txt
+```c
+pObjPresentationHints hints = nmMalloc(sizeof(ObjPresentationHints));
+if (hints == NULL) goto error_handling;
+memset(hints, 0, sizeof(ObjPresentationHints));
+```
 
-The GetFirstAttr and GetNextAttr functions each take two parameters, the inf_v pointer and the oxt* pointer, and are used to iterate through the non-mandatory attributes for the object.  GetFirstAttr should return a string naming the first attribute, and GetNextAttr should iterate through subsequent attributes.  When the attributes are exhausted, these functions should return NULL.  The attributes 'name', 'annotation', and 'content_type' should not be returned.  If the object has no other attributes, GetFirstAttr should return NULL.
+The return value, `hints : ObjPresentationHints`, contains the following useful fields which the function should set to give various useful information about the attribute.
+- `hints->Constraint : void*`: An expression for determining if a value is valid.
+- `hints->DefaultExpr : void*`: An expression defining the default value.
+- `hints->MinValue : void*`: An expression defining the minimum valid value.
+- `hints->MaxValue : void*`: An expression defining the maximum valid value.
+- `hints->EnumList : XArray`: If the attribute is a string enum, this XArray lists the valid string values.
+- `hints->EnumQuery : char*`: A query string which enumerates the valid values a string enum attribute.
+- `hints->Format : char*`: presentation format - datetime or money  <!-- TODO: Greg - We should add detail here. -->
+- `hints->AllowChars : char*`: An array of all valid characters for a string attribute, NULL to allow all characters.
+- `hints->BadChars : char*`: An array of all invalid characters for a string attribute.
+- `hints->Length : int`: The maximum length of data that can be included in a string attribute.
+- `hints->VisualLength : int`: The length that the attribute should be displayed if it is show to the user.
+- `hints->VisualLength2 : int`: The number of lines to use in a multi-line edit box for the attribute.
+- `hints->BitmaskRO : unsigned int`: which bits, if any, in bitmask are read-only <!-- TODO: Greg - Clarify which bitmask? -->
+- `hints->Style : int`: Style flags, documented below.
+- `hints->StyleMask : int`: A mask for which style flags were set and which were left unset / undefined.
+- `hints->GroupID : int`: Used to assign attributes to groups. Use -1 if the attribute is not in a group.
+- `hints->GroupName : char*`: The name of the group to which this attribute belongs, or NULL if it is ungrouped or if the group is named elsewhere.
+- `hints->OrderID : int`: Used to specify an attribute order.
+- `hints->FriendlyName : char*`: Used to specify a "display name" for an attribute (e.g. `n_rows` might have a friendly name of `"Number of Rows"`). Should be `nmSysMalloc()`ed, often using `nmSysStrdup()`.
+
+- ⚠️ **Warning**: Behavior is undefined if:
+  - If a character is included in both `hints->AllowChars` and `hints->BadChars`.
```

Comment:
BadChars should take priority.  If a char is in both, then it should be rejected.

Reply:
Updated.